File: backend/internal/models/booking.go
package models

import "time"

// BookingItem is a single line within a booking: one selected listing with
// the price snapshot at the time of the request (and guest count, for
// per-person priced venues). The whole []BookingItem slice is serialized
// as one JSON blob column (Booking.Items, `gorm:"serializer:json"`), so
// every field below is additive purely at the Go/JSON level — an old
// booking's stored JSON simply has no hall_id/menu_id keys, and decodes
// straight into these fields' zero values (nil/""/0), no migration
// needed.
//
// HallID/HallName/MenuID/MenuName/MenuPricePerGuest — restaurant/venue
// hall+menu snapshot (brief section 8). Frozen at booking time so a later
// admin edit to a ListingMenu's price never rewrites a historical
// booking's total — the same "trust and freeze the client-supplied
// snapshot" model this struct's own Name/UnitPrice/TotalPrice already use
// (nothing here is server-re-verified against the live Listing either;
// see booking_handler.go's own doc comment on why that's an intentional,
// pre-existing property of this endpoint, not something this stage
// introduces). Guests above already served as "nullable, 0 = unspecified"
// for ordinary per-guest pricing — reused as-is for a hall/menu-priced
// item's own guest count rather than adding a second, redundant field.
type BookingItem struct {
	ListingID  uint   `json:"listing_id"`
	Name       string `json:"name"`
	Category   string `json:"category"`
	Guests     uint   `json:"guests"`
	UnitPrice  uint   `json:"unit_price"`
	TotalPrice uint   `json:"total_price"`

	HallID            *uint  `json:"hall_id,omitempty"`
	HallName          string `json:"hall_name,omitempty"`
	MenuID            *uint  `json:"menu_id,omitempty"`
	MenuName          string `json:"menu_name,omitempty"`
	MenuPricePerGuest uint   `json:"menu_price_per_guest,omitempty"`

	// SelectedExtras/EstimatedTotal — the calculator's own extras
	// selection and computed total at the moment this item was added to
	// Cart (brief section 5 — added only because no equivalent field
	// existed yet). Frozen the same way as everything else above: a later
	// admin edit to a ListingMenuExtra's price never touches an already-
	// placed booking. EstimatedTotal is informational (what the customer
	// saw on the calculator, extras included) — TotalPrice above stays
	// the one figure the rest of this codebase (admin bookings list,
	// Booking.Total) already sums and displays, so nothing existing has
	// to learn about this new field to keep working.
	SelectedExtras []BookingItemExtra `json:"selected_extras,omitempty"`
	EstimatedTotal uint               `json:"estimated_total,omitempty"`
}

// BookingItemExtra is one frozen ListingMenuExtra selection — deliberately
// a tiny, flat snapshot (not a foreign key to ListingMenuExtra) for the
// same reason HallName/MenuName are plain strings and not just IDs: the
// admin row it came from can be edited or deleted later without touching
// history already written here.
type BookingItemExtra struct {
	Title string `json:"title"`
	Price uint   `json:"price"`
	Unit  string `json:"unit,omitempty"`
}

// Booking is a customer's booking request submitted from the cart — one or
// more BookingItems bundled together with contact info.
type Booking struct {
	ID        uint   `gorm:"primaryKey" json:"id"`
	PublicRef string `gorm:"size:24;uniqueIndex" json:"public_ref"`
	UserID    *uint  `gorm:"index" json:"user_id,omitempty"`
	// EventID links a Booking created from an event workspace's final
	// request (see models/event_request.go) back to that event — nil for
	// every ordinary cart checkout, which is unaffected by this column.
	EventID   *uint         `gorm:"index" json:"event_id,omitempty"`
	Name      string        `gorm:"size:150;not null" json:"name"`
	Phone     string        `gorm:"size:30;not null" json:"phone"`
	Message   string        `gorm:"type:text" json:"message"`
	Items     []BookingItem `gorm:"serializer:json" json:"items"`
	Total     uint          `json:"total"`
	Status    string        `gorm:"size:20;not null;default:new" json:"status"`
	Paid      bool          `gorm:"not null;default:false" json:"paid"`
	CreatedAt time.Time     `json:"created_at"`
	UpdatedAt time.Time     `json:"updated_at"`
}

File: flutter/lib/models/booking.dart
/// Mirrors `models.BookingItemExtra` (backend/internal/models/booking.go) —
/// the same three fields `bookingItemExtraInput` accepts, echoed back
/// unchanged on the stored/returned item.
class BookingItemExtra {
  const BookingItemExtra({
    required this.title,
    required this.price,
    required this.unit,
  });

  final String title;
  final int price;
  final String unit;

  factory BookingItemExtra.fromJson(Map<String, dynamic> json) {
    return BookingItemExtra(
      title: json['title'] as String? ?? '',
      price: json['price'] as int? ?? 0,
      unit: json['unit'] as String? ?? '',
    );
  }
}

/// Mirrors `models.BookingItem` — the exact snapshot the backend stored,
/// echoed back on the `booking` object POST /api/bookings responds with.
/// Read directly from the server's own response rather than re-derived
/// from the local cart, so the success screen/PDF show precisely what was
/// actually saved (brief section 3 — "не пересчитывать").
class BookingItem {
  const BookingItem({
    required this.listingId,
    required this.name,
    required this.category,
    required this.guests,
    required this.unitPrice,
    required this.totalPrice,
    this.hallId,
    this.hallName,
    this.menuId,
    this.menuName,
    this.menuPricePerGuest,
    this.selectedExtras = const [],
    this.estimatedTotal,
  });

  final int listingId;
  final String name;
  final String category;
  final int guests;
  final int unitPrice;
  final int totalPrice;
  final int? hallId;
  final String? hallName;
  final int? menuId;
  final String? menuName;
  final int? menuPricePerGuest;
  final List<BookingItemExtra> selectedExtras;
  final int? estimatedTotal;

  /// Same test the cart UI already uses — a hall or a menu was picked.
  bool get isRestaurantVariant => hallId != null || menuId != null;

  factory BookingItem.fromJson(Map<String, dynamic> json) {
    return BookingItem(
      listingId: json['listing_id'] as int? ?? 0,
      name: json['name'] as String? ?? '',
      category: json['category'] as String? ?? '',
      guests: json['guests'] as int? ?? 0,
      unitPrice: json['unit_price'] as int? ?? 0,
      totalPrice: json['total_price'] as int? ?? 0,
      hallId: json['hall_id'] as int?,
      hallName: (json['hall_name'] as String?)?.isEmpty ?? true
          ? null
          : json['hall_name'] as String,
      menuId: json['menu_id'] as int?,
      menuName: (json['menu_name'] as String?)?.isEmpty ?? true
          ? null
          : json['menu_name'] as String,
      menuPricePerGuest: json['menu_price_per_guest'] as int?,
      selectedExtras:
          (json['selected_extras'] as List?)
              ?.map(
                (e) => BookingItemExtra.fromJson(
                  Map<String, dynamic>.from(e as Map),
                ),
              )
              .toList() ??
          const [],
      estimatedTotal: json['estimated_total'] as int?,
    );
  }
}

/// Mirrors the `booking` object POST /api/bookings responds with
/// (backend/internal/models/booking.go's `Booking`) — the full record, not
/// just id/status, so the success screen and PDF export can be built
/// directly from what the server actually saved.
class Booking {
  const Booking({
    required this.id,
    required this.publicRef,
    required this.name,
    required this.phone,
    required this.message,
    required this.items,
    required this.total,
    required this.status,
    this.userId,
    this.eventId,
    this.createdAt,
  });

  final int id;
  final String publicRef;
  final String name;
  final String phone;
  final String message;
  final List<BookingItem> items;
  final int total;
  final String status;
  final int? userId;
  // Set for a booking created from/linked to an Event Workspace request —
  // never present for an ordinary cart checkout (see EventRequestHandler's
  // own approval flow) except via the onboarding pipeline below, which can
  // attach a brand-new event to an anonymous checkout after the fact.
  final int? eventId;
  final DateTime? createdAt;

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['id'] as int? ?? 0,
      publicRef: json['public_ref'] as String? ?? '',
      name: json['name'] as String? ?? '',
      phone: json['phone'] as String? ?? '',
      message: json['message'] as String? ?? '',
      items:
          (json['items'] as List?)
              ?.map(
                (e) =>
                    BookingItem.fromJson(Map<String, dynamic>.from(e as Map)),
              )
              .toList() ??
          const [],
      total: json['total'] as int? ?? 0,
      status: json['status'] as String? ?? 'new',
      userId: json['user_id'] as int?,
      eventId: json['event_id'] as int?,
      createdAt: DateTime.tryParse(json['created_at'] as String? ?? ''),
    );
  }
}

/// Mirrors `onboardingResult` (backend/internal/handlers/onboarding.go) —
/// present only when `BookingHandler.Create`'s best-effort booking→account
/// pipeline actually did something (brief section 14): AUTO_ACCOUNT_FROM_
/// BOOKING is on, the booking was genuinely anonymous, and it had no event
/// yet. `null` on [BookingCreateResult.onboarding] is the normal case and
/// must render nothing extra, exactly like `BookingWorkspaceCTA.jsx`'s own
/// `if (!onboarding) return null;`.
class BookingOnboarding {
  const BookingOnboarding({
    required this.status,
    this.eventId,
    this.claimToken,
    this.deliveryStatus,
    this.deliveryChannel,
  });

  /// "created_pending" | "pending_claim" | "existing_account" — never
  /// anything else; matched exactly, never guessed at.
  final String status;
  final int? eventId;
  final String? claimToken;

  /// Only ever present when a delivery attempt actually happened — see the
  /// Go struct's own doc comment on why this is `omitempty`, not a default
  /// empty string.
  final String? deliveryStatus;
  final String? deliveryChannel;

  factory BookingOnboarding.fromJson(Map<String, dynamic> json) {
    return BookingOnboarding(
      status: json['status'] as String? ?? '',
      eventId: json['event_id'] as int?,
      claimToken: json['claim_token'] as String?,
      deliveryStatus: json['delivery_status'] as String?,
      deliveryChannel: json['delivery_channel'] as String?,
    );
  }
}

/// The full `POST /api/bookings` response — `{booking, onboarding?}`.
class BookingCreateResult {
  const BookingCreateResult({required this.booking, this.onboarding});

  final Booking booking;
  final BookingOnboarding? onboarding;

  factory BookingCreateResult.fromJson(Map<String, dynamic> json) {
    return BookingCreateResult(
      booking: Booking.fromJson(
        Map<String, dynamic>.from(json['booking'] as Map),
      ),
      onboarding: json['onboarding'] == null
          ? null
          : BookingOnboarding.fromJson(
              Map<String, dynamic>.from(json['onboarding'] as Map),
            ),
    );
  }
}
